platform/duration.cpp

```diff
@@ -0,0 +1,122 @@
+
+namespace platform
+{
+namespace
+{
+using std::chrono::duration_cast, std::chrono::seconds, std::chrono::minutes, std::chrono::hours, std::chrono::days;
+
+static constexpr std::string_view kNoSpace = "";
+
+long SecondsToUnits(seconds duration, Duration::Units unit)
+{
+  switch (unit)
+  {
+    case Duration::Units::Days: return duration_cast<days>(duration).count();
+    case Duration::Units::Hours: return duration_cast<hours>(duration).count();
+    case Duration::Units::Minutes: return duration_cast<minutes>(duration).count();
+    default: UNREACHABLE();
+  }
+}
+
+seconds UnitsToSeconds(long value, Duration::Units unit)
+{
+  switch (unit)
+  {
+    case Duration::Units::Days: return days(value);
+    case Duration::Units::Hours: return hours(value);
+    case Duration::Units::Minutes: return minutes(value);
+    default: UNREACHABLE();
+  }
+}
+
+std::string_view GetUnitSeparator(Locale const &locale)
+{
+  static constexpr std::array<std::string, 12> kEmptyNumberUnitSeparatorLocales =
+  {
+    "en", "de", "fr", "he", "fa", "ja", "ko", "mr", "th", "tr", "vi", "zh"
+  };
+  bool const isEmptySeparator = std::find(std::begin(kEmptyNumberUnitSeparatorLocales), std::end(kEmptyNumberUnitSeparatorLocales), locale.m_language) != std::end(kEmptyNumberUnitSeparatorLocales);
+  return isEmptySeparator ? kNoSpace : kNarrowNonBreakingSpace;
+}
+
+std::string_view GetUnitsGroupingSeparator(Locale const & locale)
+{
+  static constexpr std::array<std::string, 2> kEmptyGroupingSeparatorLocales =
+  {
+    "ja", "zh"
+  };
+  bool const isEmptySeparator = std::find(std::begin(kEmptyGroupingSeparatorLocales), std::end(kEmptyGroupingSeparatorLocales), locale.m_language) != std::end(kEmptyGroupingSeparatorLocales);
+  return isEmptySeparator ? kNoSpace : kNonBreakingSpace;
+}
+
+bool IsUnitsOrderValid(std::initializer_list<Duration::Units> units) {
+  auto it = units.begin();
+  auto prev = *it;
+  ++it;
+  for (; it != units.end(); ++it) {
+    if (static_cast<int>(*it) <= static_cast<int>(prev))
+      return false;
+    prev = *it;
+  }
+  return true;
+}
+}
+
+Duration::Duration(long seconds) : m_seconds(seconds)
+{
+  ASSERT_GREATER_OR_EQUAL(seconds, 0, ());
+}
+
+std::string Duration::GetLocalizedString(std::initializer_list<Units> units, Locale const & locale) const
+{
+  return GetString(units, GetUnitSeparator(locale), GetUnitsGroupingSeparator(locale));
+}
+
+std::string Duration::GetPlatformLocalizedString() const
+{
+  static auto const kCurrentUnitSeparator = GetUnitSeparator(GetCurrentLocale());
+  static auto const kCurrentGroupingSeparator = GetUnitsGroupingSeparator(GetCurrentLocale());
+  return GetString({Units::Days, Units::Hours, Units::Minutes}, kCurrentUnitSeparator, kCurrentGroupingSeparator);
+}
+
+std::string Duration::GetString(std::initializer_list<Units> units, std::string_view unitSeparator, std::string_view groupingSeparator) const
+{
+  ASSERT(units.size(), ());
+  ASSERT(IsUnitsOrderValid(units), ());
```

Comment:
All such checks can be done at compile time.
```suggestion
  static_assert(units.size());
  static_assert(IsUnitsOrderValid(units));
```

Reply:
It seems like the std::initializer_list is not the constexpr variable and cannot expose its size during the compile time.
The only suggestion I found is to replace it with the std::array to use the constexpr and static_assert

Reply:
You're right, that can be done in compile time, but it will complicate the code. Let's leave it as is.